File: testsuite/tests/ada/0018-classwide/cpp/main.cpp
#include "test.h"
#include <cassert>
#include <iostream>

class CppClass : public test::Root {
public:
    CppClass(int i) : test::Root(i, this) {}

    CppClass(const CppClass&, gnatpolyglot::data *data)
        : test::Root(data) {}

    CppClass *internal_clone_(gnatpolyglot::data *data) {
        return new CppClass(*this, data);
    }

    gnatpolyglot::polyglot_ptr<test::Root>
    f_root(const test::Root &r2) const override {
        std::cout << "f_root: "
                  << (dynamic_cast<const ::test::Child *>(&r2) != nullptr)
                  << "\n";
        return gnatpolyglot::polyglot_ptr<test::Root>(new test::Root(r2));
    }

    void p_root_a(gnatpolyglot::polyglot_ptr<test::Root> r2) const override {
        std::cout << "p_root_a: "
                  << (dynamic_cast<const ::test::Child *>(r2.get()) != nullptr)
                  << "\n";
    }

    void p_out_root_a(gnatpolyglot::polyglot_ptr<test::Root> &r2) const override {
        std::cout << "p_out_root_a: "
                  << (dynamic_cast<const ::test::Child *>(r2.get()) != nullptr)
                  << "\n";
    }
};

int main() {
    assert(dynamic_cast<::test::Child*>(test::get_root().get()) == nullptr);
    assert(dynamic_cast<::test::Child*>(test::get_child().get()) != nullptr);

    gnatpolyglot::polyglot_ptr<test::Root> ptr1 = test::get_root_a();
    assert(dynamic_cast<::test::Child*>(ptr1.get()) == nullptr);
    ptr1.set_owner(gnatpolyglot::memory_owner::USER);
    gnatpolyglot::polyglot_ptr<test::Root> ptr2 = test::get_child_a();
    assert(dynamic_cast<::test::Child*>(ptr2.get()) != nullptr);
    ptr2.set_owner(gnatpolyglot::memory_owner::USER);

    {
        gnatpolyglot::polyglot_ptr<test::Root> ptr(nullptr);
        test::get_root(ptr);
        assert(ptr);
        assert(dynamic_cast<::test::Child*>(ptr.get()) == nullptr);
        test::get_child(ptr);
        assert(dynamic_cast<::test::Child*>(ptr.get()) != nullptr);
        ptr.set_owner(gnatpolyglot::memory_owner::USER);
    }
    {
        CppClass c{4};
        test::call_f_root(c, test::Root(1));
        test::call_f_root(c, test::Child(1, 2));
        test::Root tmp_r{1};
        test::call_p_root_a(c, gnatpolyglot::polyglot_ptr<test::Root>{tmp_r});
        gnatpolyglot::polyglot_ptr<test::Root> tmp_ptr1{tmp_r};
        test::call_p_out_root_a(c, tmp_ptr1);
        test::Child tmp_c{1, 2};
        test::call_p_root_a(c, gnatpolyglot::polyglot_ptr<test::Root>{tmp_c});
        gnatpolyglot::polyglot_ptr<test::Root> tmp_ptr2{tmp_c};
        test::call_p_out_root_a(c, tmp_ptr2);

        gnatpolyglot::polyglot_ptr<test::Root> cloned = test::clone(c);
        assert(dynamic_cast<CppClass*>(cloned.get()) != nullptr);
        cloned->set_i(5);
        std::cout << c.get_i() << " != " << (cloned->get_i()) << "\n";

        gnatpolyglot::polyglot_ptr<test::Root> c2 {new CppClass(5)};
        test::free(c2);
        std::cout << c2.get() << "\n";
    }
    {
        CppClass c(2);
        auto f = [](const test::Root& r) {
            std::cout << "call_callback: "
                      << (dynamic_cast<const ::test::Child *>(&r) != nullptr)
                      << (dynamic_cast<const CppClass *>(&r) != nullptr)
                      << "\n";
        };
        test::call_callback(f, test::Root(1));
        test::call_callback(f, test::Child(1, 2));
        test::call_callback(f, c);
    }
}

File: testsuite/tests/ada/0018-classwide/java/Main.java
import com.adacore.libtest.Callbacks.Callback0;
import com.adacore.libtest.test.Child;
import com.adacore.libtest.test.Root;
import com.adacore.libtest.test.TestPackage;

import com.adacore.gnatpolyglot.runtime.PolyglotData.Owner;

public class Main {

    static class JavaChild extends Root {
        public JavaChild(int i) {
            super(i);
        }

        @Override
        public java.util.Optional<Root> fRoot(Root r2) {
            System.out.println("pRoot: " + (r2 instanceof Child));
            return java.util.Optional.of(new Root(r2));
        }

        @Override
        public void pRootA(Root r2) {
            System.out.println("pRootA: " + (r2 instanceof Child));
        }

        @Override
        public void pOutRootA(Root.Ref r2) {
            System.out.println("pOutRootA: " + (r2.get().get() instanceof Child));
        }

    }

    public static void main(String[] args) throws Throwable {
        assert TestPackage.getRoot().map(v -> !(v instanceof Child)).orElse(false);
        assert TestPackage.getChild().map(Child.class::isInstance).orElse(false);

        assert TestPackage.getRootA().map(v -> !(v instanceof Child)).orElse(false);
        assert TestPackage.getChildA().map(Child.class::isInstance).orElse(false);

        Root.Ref ptr = new Root.Ref(null);
        TestPackage.getRoot(ptr);
        assert ptr.get().isPresent();
        assert ptr.get().map(Object::getClass).map(Root.class::equals).orElse(false);
        TestPackage.getChild(ptr);
        assert ptr.get().map(Object::getClass).map(Child.class::equals).orElse(false);

        JavaChild c = new JavaChild(4);
        TestPackage.callFRoot(c, new Root(1));
        TestPackage.callFRoot(c, new Child(1, 2));

        Root tmpR = new Root(1);
        tmpR._setOwner(Owner.LIBRARY);
        TestPackage.callPRootA(c, tmpR);
        TestPackage.callPOutRootA(c, new Root.Ref(tmpR));
        Child tmpC = new Child(1, 2);
        tmpC._setOwner(Owner.LIBRARY);
        TestPackage.callPRootA(c, tmpC);
        TestPackage.callPOutRootA(c, new Root.Ref(tmpC));
        tmpR._setOwner(Owner.USER);
        tmpC._setOwner(Owner.USER);

        java.util.Optional<Root> cloned = TestPackage.clone(c);
        assert cloned.map(JavaChild.class::isInstance).orElse(false);
        cloned.ifPresent(r -> r.setI(5));
        System.out.println(
            c.getI() + " != " + cloned.map(Root::getI).get()
        );

        Root.Ref c2 = new Root.Ref(new JavaChild(5));
        c2.get().ifPresent(root -> root._setOwner(Owner.LIBRARY));
        TestPackage.free(c2);
        System.out.println(c2.get());

        Callback0 cb = (r) -> {
            System.out.format(
                "call_callback: %s, %s\n",
                r instanceof Child,
                r instanceof JavaChild
            );
        };
        TestPackage.callCallback(cb, new Root(1));
        TestPackage.callCallback(cb, new Child(1, 2));
        TestPackage.callCallback(cb, new JavaChild(2));
    }
}
